edgexfoundry/app-service-configurable
# feat(snap): add support for config interface and new secrets import file

Signed-off-by: Mengyi <mengyi.wang@canonical.com>







**If your build fails**  due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/app-service-configurable/blob/main/.github/CONTRIBUTING.md

## PR Checklist
Please check if your PR fulfills the following requirements:


## Testing Instructions

```
git clone https://github.com/MonicaisHer/app-service-configurable.git
cd app-service-configurable
git checkout add-support-for-new-secrets-import-file

snapcraft

snap remove --purge edgex-app-service-configurable
snap install ./edgex-app-service-configurable_2.0.2-dev.30_amd64.snap --dangerous
snap remove --purge edgexfoundry
snap install edgexfoundry --edge
```
test 1 - negative test: validate that edgex-app-service-configurable can not find test.json, it means config hook works by passing value:
```
sudo snap set edgex-app-service-configurable profile=mqtt-export
sudo install -vDT \
/var/snap/edgexfoundry/current/secrets/app-mqtt-export/secrets-token.json \
/var/snap/edgex-app-service-configurable/current/app-mqtt-export/secrets-token.json
sudo snap set edgex-app-service-configurable env.secret-store.secrets-file=test.json 
sudo snap start edgex-app-service-configurable
sudo snap logs edgex-app-service-configurable -n 100 | grep test.json
```
expect:
2021-11-17T16:23:45Z edgex-app-service-configurable.app-service-configurable[86789]: level=INFO ts=2021-11-17T16:23:45.407466894Z app=app-mqtt-export source=variables.go:352 msg="Variables override of 'SecretStore.SecretsFile' by environment variable: SECRETSTORE_SECRETSFILE=test.json"
2021-11-17T16:23:45Z edgex-app-service-configurable.app-service-configurable[86789]: level=ERROR ts=2021-11-17T16:23:45.411257077Z app=app-mqtt-export source=bootstrap.go:46 msg="failed to create SecretProvider: seeding secrets failed: open test.json: no such file or directory"

test 2 - validate ecret-store.secrets-file has been seeded and edgex-app-service-configurable works:
```
sudo nano /var/snap/edgex-app-service-configurable/current/test.json
add:
{
    "secrets": [
        {
            "path": "credentials001",
            "imported": false,
            "secretData": [
                {
                    "key": "username",
                    "value": "my-user-1"
                },
                                {
                    "key": "password",
                    "value": "password-001"
                }
            ]
        }
    ]
}
sudo snap set edgex-app-service-configurable env.secret-store.secrets-file="/var/snap/edgex-app-service-configurable/current/test.json"
sudo snap start --enable edgex-app-service-configurable.app-service-configurable 
sudo snap logs edgex-app-service-configurable -n 100 
```
expect:
2021-11-17T16:34:41Z edgex-app-service-configurable.app-service-configurable[87671]: level=INFO ts=2021-11-17T16:34:41.7164932Z app=app-mqtt-export source=variables.go:352 msg="Variables override of 'SecretStore.SecretsFile' by environment variable: SECRETSTORE_SECRETSFILE=/var/snap/edgex-app-service-configurable/current/test.json"
[...]
2021-11-17T16:34:41Z edgex-app-service-configurable.app-service-configurable[87671]: level=INFO ts=2021-11-17T16:34:41.734713597Z app=app-mqtt-export source=secure.go:289 msg="Seeding 1 Service Secrets"
[...]
2021-11-17T16:34:41Z edgex-app-service-configurable.app-service-configurable[87671]: level=INFO ts=2021-11-17T16:34:41.738991676Z app=app-mqtt-export source=secure.go:278 msg="Scrubbing of secrets file complete."
[...]
2021-11-17T16:34:42Z edgex-app-service-configurable.app-service-configurable[87671]: level=INFO ts=2021-11-17T16:34:42.555114049Z app=app-mqtt-export source=service.go:184 msg="app-mqtt-export has Started"
2021-11-17T16:34:42Z edgex-app-service-configurable.app-service-configurable[87671]: level=INFO ts=2021-11-17T16:34:42.555153764Z app=app-mqtt-export source=server.go:156 msg="Starting HTTP Web Server on address localhost:59703"
2021-11-17T16:34:42Z edgex-app-service-configurable.app-service-configurable[87671]: level=INFO ts=2021-11-17T16:34:42.555127655Z app=app-mqtt-export source=messaging.go:118 msg="Waiting for messages from the MessageBus on the 'edgex/events/#' topic"

test 3 - validate that when disable-scrub-secrets-file is set, secret-store.secrets-file will not overwrite the default secret:
```
snap remove --purge edgex-app-service-configurable
snap install ./edgex-app-service-configurable_2.0.2-dev.30_amd64.snap --dangerous
snap remove --purge edgexfoundry
snap install edgexfoundry --edge

sudo snap set edgex-app-service-configurable profile=mqtt-export
sudo snap set edgex-app-service-configurable env.secret-store.disable-scrub-secrets-file=true
sudo nano /var/snap/edgex-app-service-configurable/current/test.json
add:
{
    "secrets": [
        {
            "path": "credentials001",
            "imported": false,
            "secretData": [
                {
                    "key": "username",
                    "value": "my-user-1"
                },
                                {
                    "key": "password",
                    "value": "password-001"
                }
            ]
        }
    ]
}

sudo snap set edgex-app-service-configurable env.secret-store.secrets-file="/var/snap/edgex-app-service-configurable/current/test.json"
sudo snap start --enable edgex-app-service-configurable.app-service-configurable 
sudo snap logs edgex-app-service-configurable
```
expect:
2021-11-18T07:32:11Z edgex-app-service-configurable.app-service-configurable[106614]: level=WARN ts=2021-11-18T07:32:11.740562019Z app=app-mqtt-export source=secret.go:96 msg="Retryable failure while creating SecretClient: open /var/snap/edgex-app-service-configurable/x1/app-mqtt-export/secrets-token.json: no such file or directory"
## New Dependency Instructions (If applicable)
